### PR TITLE
Ensure deferred functions are executed before Go runtime exits

### PR DESCRIPTION
**Pull request by manstis**
## Description
Fixes https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1314#issuecomment-1322325692

## Verification Steps
1. Run integration tests twice without tearing down the database.

## Checklist (Definition of Done)
- [ ] ~All acceptance criteria specified in JIRA have been completed~
- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~Documentation added for the feature~
- [ ] CI and all relevant tests are passing
- [x] Code Review completed
- [x] Verified independently by reviewer
- [x] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] ~Required metrics/dashboards/alerts have been added (or PR created).~
- [ ] ~Required Standard Operating Procedure (SOP) is added.~
- [ ] ~JIRA has been created for changes required on the client side~

### CODE REVIEW

**Comment by machi1990** [comment]
@miguelsorianod can you've a look at this? Seems related to #1314

File: internal/connector/test/integration/feature_test.go [hunk 1]
```diff
@@ -82,9 +83,10 @@ func TestMain(m *testing.M) {
 		},
 		connector.ConfigProviders(false),
 	)
+	defer func() { os.Exit(exitCode) }()
 	defer teardown()
```
Comment (manstis) [inline internal/connector/test/integration/feature_test.go:87]: This deferred method was not being called following `os.Exit(..)`.

If you run `make db/login` after running the integration tests you can see the database tables and content are still there... they should have been removed by this `teardown()` method invocation.

Furthermore, you can also verify this by placing the following in `internal/connector/internal/migrations/migrations.go`
```
...
addOrgIDAnnotations("202212050000"),
	&gormigrate.Migration{
		ID: "test",
		Migrate: func(g *gorm.DB) error {
			fmt.Println("============== RUNNING MIGRATION")
			return nil
		},
		Rollback: func(g *gorm.DB) error {
			fmt.Println("============== RUNNING ROLLBACK")
			return nil
		},
...
```
You can observe that the rollback method is not called.

File: internal/kafka/test/integration/kafka/kafka_test.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ func init() {
 }
 
 func TestMain(m *testing.M) {
+	exitCode := 0
```
Comment (manstis) [inline internal/kafka/test/integration/kafka/kafka_test.go:25]: I made a similar change for `kas` as it has the same problem (but it does not manifest itself).

File: internal/connector/test/integration/feature_test.go (hunk 1 above)
Comment (manstis) [reply internal/connector/test/integration/feature_test.go:87]: I followed the approach suggested [here](https://www.golinuxcloud.com/golang-handle-defer-calls-os-exit/) ("Method-2")

File: test/helper.go [hunk 3]
```diff
@@ -141,16 +141,16 @@ func NewHelperWithHooksAndDBsetup(t *testing.T, httpServer *httptest.Server, set
 		glog.Fatalf("Unable to initialize testing environment: %s", err.Error())
 	}
 
-	h.CleanDB()
```
Comment (manstis) [inline test/helper.go:144]: `CleanDB()` is [called by](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/test/helper.go#L334) `ResetDB()` (on L#145). No need to call it twice.

File: test/helper.go [hunk 4]
```diff
@@ -141,16 +141,16 @@ func NewHelperWithHooksAndDBsetup(t *testing.T, httpServer *httptest.Server, set
...
 	h.SetupDB(setupDBsqlStatements)
 	env.Start()
 	return h, buildTeardownHelperFn(
-		env.Stop,
 		h.CleanDB,
 		metrics.Reset,
 		stopJWKMockServer,
-		env.Cleanup)
+		env.Cleanup,
+		env.Stop,
```
Comment (manstis) [inline test/helper.go:152]: It made more sense to me to "stop" the environment (which asks each `BootService` to _stop_) last.

The code also asks each `BootService` to "clean up" which, IMO, should happen before they stop).

Happy to revert the order of these calls; if others disagree.

File: test/helper.go (hunk 3 above)
Comment (miguelsorianod) [reply test/helper.go:144]: :+1:

File: internal/kafka/test/integration/kafka/kafka_test.go [hunk 5]
```diff
@@ -41,9 +42,10 @@ func TestMain(m *testing.M) {
 	defer ocmServer.Close()
 
 	helper, _, teardown = ktest.NewKafkaHelper(t, ocmServer)
+	defer func() { os.Exit(exitCode) }()
```
Comment (miguelsorianod) [inline internal/kafka/test/integration/kafka/kafka_test.go:45]: Shouldn't this be moved above the defer call to `ocmServer.Close()` ?

The order on which deferred calls are run is LIFO and I understand the Exit method call should be performed at the end.

File: internal/connector/test/integration/feature_test.go (hunk 1 above)
Comment (miguelsorianod) [reply internal/connector/test/integration/feature_test.go:87]: Good catch :+1: .

The reason about why deferred was not being called is documented in the `os.Exit` function documentation available in https://pkg.go.dev/os#Exit:

Exit causes the current program to exit with the given status code. Conventionally, code zero indicates success, non-zero an error. The program terminates immediately; **deferred functions are not run**. 

The approach used looks good to me :+1:

File: test/helper.go (hunk 4 above)
Comment (manstis) [reply test/helper.go]: I've restored the clean-up order.. 

Thinking about it over the weekend `env.Cleanup` should probably be last; after everything has stopped.

File: internal/kafka/test/integration/kafka/kafka_test.go (hunk 5 above)
Comment (manstis) [reply internal/kafka/test/integration/kafka/kafka_test.go:45]: Thanks. Good spot. I didn't see the other deferral. Fixed.

**Review by miguelsorianod — CHANGES_REQUESTED** [review]
The changes look good to me :+1: 

Could you squash redundant commits to end up with a clean history?

After that and the tests pass it will be good to merge :+1:

**Comment by manstis** [comment]
@miguelsorianod 

> Could you squash redundant commits to end up with a clean history?

Squashed.

**Comment by miguelsorianod** [comment]
It seems the action times out @manstis . I see it is currently set to `14` minutes.
It is interesting that this is happening in this PR. The changes here should not have substantially increased the execution time. Maybe it was already at the limit and successfully running the teardown function ends up going over the thresold.

We can increase the timeout of the action.

**Comment by manstis** [comment]
@miguelsorianod 

> We can increase the timeout of the action and see what happens.

Quite probably. Do you know where/who/how it can be done? I'm not an admin on this GitHub repo.

**Comment by manstis** [comment]
@miguelsorianod Please see https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1632

That will need merging before it affects this PR.. 

There are some "tricks" that can be used; but since it's only a "timeout" property merging it IMO should is OK.

**Comment by manstis** [comment]
@miguelsorianod As @machi1990 pointed out there are 2 timeouts defined:
1. GitAction Workflow configuration
2. `gotestsum` configuration

Can we be certain which has caused the timeout?

**Comment by miguelsorianod** [comment]
> @miguelsorianod As @machi1990 pointed out there are 2 timeouts defined:
> 
>     1. GitAction Workflow configuration
> 
>     2. `gotestsum` configuration
> 
> 
> Can we be certain which has caused the timeout?
> 
> I think we should align both at 20 minutes and, if it still fails, investigate again.

I understand the action itself caused the timeout, as shown in the logs:
```
Error: The action has timed out.
```

**Comment by manstis** [comment]
If the _action_ was the integration test it'd have been timed out by `gotestsum` at 14 minutes..

So again, I propose the PR aligning `gotestsum` timeout to 20 minutes is merged and then we re-assess.

**Comment by miguelsorianod** [comment]
> If the _action_ was the integration test it'd have been timed out by `gotestsum` at 14 minutes..
> 
> So again, I propose the PR aligning `gotestsum` timeout to 20 minutes is merged and then we re-assess.

That's not correct. The `gotestsum` timeout has a timeout, but `gotestsum` is run several times. One time for kafka tests and one time for connectors tests. So it can go over the action timeout.

In any case, if you really want to match the timeouts you can add a temporary commit in this PR.

**Comment by manstis** [comment]
Hi @machi1990 @miguelsorianod I found an issue in COS where a web server connection was not being closed.

This lead to a failure to stop the API Server in integration tests and the timeout (now we try to _teadown_).

It looks like a [change](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/commit/2db1ce1bd2ecc9fb7bb558eb372474358917e94d) in KAS lead to the undesirable scenario.

All good now though.

Shall I rollback the CI timeout change?

**Comment by manstis** [comment]
@lburgazzoli @valdar @dhirajsb [FYI](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1625#issuecomment-1467998563).

I was told support for "watch" was to be removed for Connectors (and hence not to be supported for Processors).

Was this based on there being issues with its operation (now fixed) or otherwise?

I'm just wondering whether support now needs to be removed.

**Comment by miguelsorianod** [comment]
> Shall I rollback the CI timeout change?

If the increase of the observed running times was due to the described connectors issue then yes the timeout change should be rolled back

**Comment by manstis** [comment]
> ...  the timeout change should be rolled back

See https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1636

**Comment by miguelsorianod** [comment]
> @lburgazzoli @valdar @dhirajsb [FYI](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1625#issuecomment-1467998563).
> 
> I was told support for "watch" was to be removed for Connectors (and hence not to be supported for Processors).
> 
> Was this based on there being issues with its operation (now fixed) or otherwise?
> 
> I'm just wondering whether support now needs to be removed.

@manstis do you want to wait to get an answer to this before proceeding with the PR merge or should we go ahead?

**Review by miguelsorianod — APPROVED** [review]

**Pull request merged**